system/security/shared/secmanagertracedecorator.hpp

```diff
@@ -0,0 +1,364 @@
+    { \
+        spanScope->recordException(e); \
+        throw; \
+    } \
+    catch (...) \
+    { \
+        Owned<IException> e(makeStringException(-1, "unknown exception")); \
+        spanScope->recordException(e); \
+        throw; \
+    }
+
+/**
+ * @brief Decorator for ISecManager that adds tracing to the interface.
+ *
+ * Tracing is added to selected methods of ISecManager. Traced methods create a span for the
+ * duration of the decorated method. Untraced methods are passed through to the decorated object.
+ *
+ * The default decorated type is sufficient for most security managers. CLdapSecManager is an
+ * exception because platform processes depend on interfaces not included in the default.
+ * - If ISecManager and CLdapSecManager interfaces are standardized, this point becomes moot.
+ * - If ISecManager is extended to declare LDAP-specific interfaces, and CLdapSecManager implements
+ *   this interface, create a subclass of TSecManagerTraceDecorator to decorate the new interface
+ *   and replace ISecManager with the new interface name as template parameters.
+ * - If nothing else changes, create a subclasses of TSecManagerTraceDecorator, with template
+ *   parameters CLdapSecManager and ISecManager, to decorate the LDAP-specific interfaces.
+ */
+template <typename secmgr_t = ISecManager, typename secmgr_interface_t = ISecManager>
+class TSecManagerTraceDecorator : public TDecorator<secmgr_t, secmgr_interface_t>
+{
+    using TDecorator<secmgr_t, secmgr_interface_t>::decorated;
+public:
+    virtual SecFeatureSet queryFeatures(SecFeatureSupportLevel level) const override
+    {
+        return decorated->queryFeatures(level);
+    }
+    virtual ISecUser * createUser(const char * user_name, IEspSecureContext* secureContext = nullptr) override
+    {
+        return decorated->createUser(user_name, secureContext);
+    }
+    virtual ISecResourceList * createResourceList(const char * rlname, IEspSecureContext* secureContext = nullptr) override
+    {
+        return decorated->createResourceList(rlname, secureContext);
+    }
+    virtual bool subscribe(ISecAuthenticEvents & events, IEspSecureContext* secureContext = nullptr) override
+    {
+        return decorated->subscribe(events, secureContext);
+    }
+    virtual bool unsubscribe(ISecAuthenticEvents & events, IEspSecureContext* secureContext = nullptr) override
+    {
+        return decorated->unsubscribe(events, secureContext);
+    }
+    virtual bool authorize(ISecUser & user, ISecResourceList * resources, IEspSecureContext* secureContext = nullptr) override
+    {
+        START_SEC_MANAGER_TRACE_BLOCK("security.authorize");
+            return decorated->authorize(user, resources, secureContext);
+        END_SEC_MANAGER_TRACE_BLOCK
+    }
+    virtual bool authorizeEx(SecResourceType rtype, ISecUser & user, ISecResourceList * resources, IEspSecureContext* secureContext = nullptr) override
+    {
+        START_SEC_MANAGER_TRACE_BLOCK("security.authorize_ex");
```

Comment:
Should this be "security.authorize_ex_list" to distinguish between a single authorization and a list? It would seem that ia a long list was passed that it could skew results if it's measured the same as a single authorization. There is at least one other case where a method is overloaded to accept both a single instance and a list.

Reply:
It can be. I thought about it because it's already a distinction made in the feature flags.